src/Extend/User.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Extend;
+
+use Flarum\Extension\Extension;
+use Illuminate\Contracts\Container\Container;
+
+class User implements ExtenderInterface
+{
+    private $groupProcessors = [];
+
+    /**
+     * Dynamically process a user's list of groups when calculating permissions.
+     * This can be used to give a user permissions for groups they aren't actually in, based on context.
+     * It will not change the group badges displayed for the user.
+     *
+     * @param callable $callable
+     *
+     * The callable can be a closure or invokable class, and should accept:
+     * - \Flarum\User\User $user: the user in question.
+     * - array $groupIds: an array of ids for the groups the user belongs to.
+     *
+     * The callable should return:
+     * - array $groupIds: an array of ids for the groups the user belongs to.
+     */
+    public function addGroupProcessor(callable $callable)
```

Comment:
`addGroupProcessor` again sounds awkwardly close to our internal implementation details. What about just `groups()`? Since we're passing a callback anyway, this allows for all kinds of manipulation, so we're golden.

Since the callbacks are not named, there isn't going to be any removal anyway, so by convention we wouldn't use the `add()` prefix here.

Reply:
I just read the docblock, ahem. :see_no_evil: 

Since this is just for calculating permissions, I guess `groups()` would be a bit too generic and could confuse developers. Hmm... `permissionGroups()`? Other ideas?